src/raftstore/coprocessor/split_check/size.rs

```diff
@@ -16,38 +16,53 @@ use rocksdb::DB;
 use util::transport::{RetryableSendCh, Sender};
 
 use super::super::metrics::*;
-use super::super::{Coprocessor, ObserverContext, SplitCheckObserver, SplitChecker};
+use super::super::{Coprocessor, ObserverContext, RowEntry, SplitCheckObserver, SplitChecker};
 use super::Host;
 
 pub struct Checker {
     max_size: u64,
     split_size: u64,
+    max_rows: u64,
+    split_rows: u64,
     current_size: u64,
+    current_rows: u64,
     split_key: Option<Vec<u8>>,
 }
 
 impl Checker {
-    pub fn new(max_size: u64, split_size: u64) -> Checker {
+    pub fn new(max_size: u64, split_size: u64, max_rows: u64, split_rows: u64) -> Checker {
         Checker {
             max_size,
             split_size,
+            max_rows,
+            split_rows,
             current_size: 0,
+            current_rows: 0,
             split_key: None,
         }
     }
+
+    fn should_split(&self) -> bool {
+        self.current_size >= self.max_size || self.current_rows >= self.max_rows
+    }
 }
 
 impl SplitChecker for Checker {
-    fn on_kv(&mut self, _: &mut ObserverContext, key: &[u8], value_size: u64) -> bool {
-        self.current_size += key.len() as u64 + value_size;
-        if self.current_size > self.split_size && self.split_key.is_none() {
-            self.split_key = Some(key.to_vec());
+    fn on_kv(&mut self, _: &mut ObserverContext, row: &RowEntry) -> bool {
+        self.current_size += row.key().len() as u64 + row.value_size() as u64;
+        if row.is_from_write_cf() {
```

Comment:
It is actually a "KeyEntry", so you are calculating the number of keys here, not rows.